File: src/seeborg4.py
import logging
import random


class SeeBorg4:
    def __init__(self, client, config, database):
        """
        :param client: ``discord.Client``
        :param config:
        :param database:
        """
        self.__logger = logging.getLogger(SeeBorg4.__name__)
        self.__client = client
        self.__config = config

    def start(self):
        self.__logger.info('SeeBorg4 is starting')
        self.__register_listeners()
        self.__client.run(self.__config.token())

    async def on_ready(self):
        """
        Event listener for on_ready event.
        Logs to the console informing the user that the client is connected.
        """
        self.__logger.info('Connected to Discord!')

    async def on_message(self, message):
        """
        Event listener for on_message event.

        When a message is received, log it to the console and then decide if
        the message should be processed by the learn/speak logic.

        :param message: ``discord.Message``
        """
        self.__logger.info('MSG %s %s: %s' % (
            message.channel.id, message.author.id, message.content))

        if not self.__should_process(message):
            return

        if self.__should_learn(message):
            self.__learn(message.clean_content)

        if self.__should_reply(message):
            self.__reply(message.channel, message.clean_content)

    def __should_process(self, message):
        """
        Check if the given message should be processed by SeeBorg4.

        A message should be processed if all conditions are met:
            - The message doesn't belong to our own client
            - The author of the message isn't in the ignore list

        :param message: ``discord.Message``
        :return: ``bool``
        """
        # Ignore own messages
        if self.__is_own_message(message):
            return False

        # Ignore users in ignore list
        if self.__config.is_ignored(message.author.id, message.channel.id):
            self.__logger.debug('IGNORE %s' % message.author.id)
            return False

        return True

    def __register_listeners(self):
        """
        Registers listeners to our client.
        """

        @self.__client.event
        async def on_ready():
            await self.on_ready()

        @self.__client.event
        async def on_message(message):
            await self.on_message(message)

    def __should_learn(self, message):
        """
        Check if the given message should be learned by SeeBorg4.

        A message will be learned if all following conditions are met:
            - Learning is enabled in the configuration
            - The contents of the message don't match any blacklisted patterns

        :param message: ``discord.Message``
        :return: ``bool``
        """
        if not self.__config.learning(message.channel.id):
            return False

        # Ignore messages that match the blacklist
        if self.__config.matches_blacklisted_pattern(message.channel.id,
                                                     message.clean_content):
            self.__logger.debug('IS BLACKLISTED %s' % message)
            return False

        return True

    def __should_reply(self, message):
        """
        Check if the given message should be replied to by SeeBorg4.

        A message is eligible for a response if all following conditions are
        met:
            - Speaking is enabled in our configuration
            - Any of the following conditions are met:

                - The message mentions our user and our random number conforms
                  to our chance of replying to a mention
                - The message's content matches a magic word pattern and our
                  random number conforms to our chance of replying to a magic
                  sentence
                - Our random number conforms to our chance of replying to a
                  regular sentence

        :param message: ``discord.Message``
        :return: ``bool``
        """
        if not self.__config.speaking(message.channel.id):
            return False

        def chance_predicate(chance_percentage, predicate):
            randint = random.randint(0, 99)  # Generate a number between 0 and 99 (inclusive)
            if chance_percentage > 0 and predicate():
                if chance_percentage > randint or chance_percentage == 100:
                    return True
            return False

        reply_mention = self.__config.reply_mention(message.channel.id)

        if chance_predicate(reply_mention, lambda: self.__client.user in message.mentions):
            self.__logger.debug('REPLY MENTION')
            return True

        reply_magic = self.__config.reply_magic(message.channel.id)

        def has_magic_pattern():
            return self.__config.matches_magic_pattern(message.channel.id, message.clean_content)

        if chance_predicate(reply_magic, lambda: has_magic_pattern()):
            self.__logger.debug('REPLY MAGIC')
            return True

        reply_rate = self.__config.reply_rate(message.channel.id)

        if chance_predicate(reply_rate, lambda: True):
            self.__logger.debug('REPLY RATE')
            return True

        self.__logger.debug('REPLY FAIL')
        return False

    def __is_own_message(self, message):
        """
        Returns true if the specified message belongs to our `_client`.

        :param message: ``discord.Message``
        :return: ``bool``
        """
        return message.author.id == self.__client.user.id

    def __learn(self, line):
        """
        #TODO

        :param line ``str``
        """
        self.__logger.debug('In method: __learn')

    def __reply(self, channel, line):
        """
        #TODO
        Builds an answer to the given line and sends the answer to the given
        channel.

        :param channel: ``discord.Channel``
        :param line ``str``
        """
        self.__logger.debug('In method: __reply')

File: src/seeborg4.js
"use strict";
const assert = require("assert");

const confmod = require("./confmod");
const logger = require("./logging").getLogger(module);
const {
  PluginManager
} = require("./pluginman");
const {
  CommandHandler
} = require("./cmdhnd");
const {
  VoiceHandler
} = require("./voicehnd");
const {
  Learner
} = require("./learn");
const {
  Answerer
} = require("./answerer");

const instances = new Set();

/**
 * Destroys all SeeBorg4 instances.
 *
 * @returns {void}
 */
function cleanup() {
  for (let instance of instances) {
    instance.destroy();
  }
  assert(instances.size === 0);
}

class SeeBorg4 {
  /**
   * Creates an instance of SeeBorg4.
   *
   * @param {*} client The chat client to use
   * @param {*} config The configuration file to use
   * @param {*} database The database to use
   * @memberof SeeBorg4
   *
   * @prop {*} client
   * @prop {*} config
   * @prop {*} database
   *
   * @prop {PluginManager} pluginManager
   * @prop {CommandHandler} commandHandler
   * @prop {VoiceHandler} voiceHandler
   * @prop {Learner} learner
   * @prop {Answerer} answerer
   *
   * @prop {?Number} autoSaveJobId ID of the JavaScript timer running the autosave job
   */
  constructor(client, config, database) {
    instances.add(this);
    this.client = client;
    this.config = config;
    this.database = database;

    this.pluginManager = new PluginManager(this);
    this.commandHandler = new CommandHandler(this);
    this.voiceHandler = new VoiceHandler(this);
    this.learner = new Learner(this);
    this.answerer = new Answerer(this);

    this.autoSaveJobId = null;
  }

  start() {
    logger.info("SeeBorg4 is starting!");
    logger.info("Loading plugins.");
    this.pluginManager.loadPlugins();
    logger.info(`Loaded ${this.pluginManager.loadedPlugins.length} plugins.`);

    logger.info("Registering listeners.");
    this.registerListeners();

    logger.info("Starting auto save job.");
    this.startAutoSaveJob();

    logger.info("Logging in!");
    this.client.login(this.config.token)
      .catch(err => {
        logger.error("Failed to log in to Discord. " + err);
      });
  }

  /**
   * Destroys this instance and makes it unusable.
   * @returns {void}
   */
  destroy() {
    logger.info("Destroy: Shutting off plugins.");
    this.pluginManager.destroy();

    logger.info("Destroy: Shutting down the client.");
    this.client.destroy();

    logger.info("Destroy: Stopping auto save job.");
    clearInterval(this.autoSaveJobId);

    logger.info("Destroy: Saving before quitting.");
    this.database.save();

    logger.info("Destroy: Removing bot from list of instances.");
    assert(instances.delete(this));

    logger.info("Destroy: Done.");
  }

  registerListeners() {
    this.client.on("ready", this.onReady.bind(this));
    this.client.on("message", this.onMessage.bind(this));
  }

  startAutoSaveJob() {
    this.autoSaveJobId = setInterval(() => {
      logger.info("Saving dictionary...");
      this.database.save();
      logger.info("Saved!");
    }, this.config.autoSavePeriod * 1000);
  }

  onReady() {
    logger.info("Connected to Discord!");
  }

  onMessage(message) {
    logger.info(`Message in #${message.channel.name} (${message.channel.id}): ${message.author.username} (${message.author.id}) >>> ${message.content}`);

    const handled = this.commandHandler.handle(message);
    if (handled) {
      logger.debug("Message handled by command handler.");
      return;
    }
    this.answerer.apply(message);
    this.learner.apply(message);
  }

  /**
   * Returns true if the user is ignored in the given channel.
   *
   * @param {*} user The user
   * @param {*} channel The channel
   * @returns {boolean}
   * @memberof SeeBorg4
   */
  isIgnored(user, channel) {
    // Ignore own messages
    if (user.id === this.client.user.id) {
      return true;
    }

    // Ignore users in the ignore list
    if (confmod.isIgnored(this.config, user.id, channel.id)) {
      return true;
    }

    return false;
  }
}

module.exports = {
  instances: instances,
  cleanup: cleanup,
  SeeBorg4: SeeBorg4
};